File: src/aiperf/config/mlflow.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""
AIPerf Configuration v2.0 - Pydantic Models

MLflow - Tracking and artifact-upload configuration.
"""

from __future__ import annotations

from typing import Annotated

from pydantic import ConfigDict, Field

from aiperf.config.base import BaseConfig
from aiperf.config.defaults import MLflowDefaults

__all__ = [
    "MLflowConfig",
]


class MLflowConfig(BaseConfig):
    """MLflow tracking and artifact-upload configuration."""

    model_config = ConfigDict(extra="forbid", validate_default=True)

    tracking_uri: Annotated[
        str | None,
        Field(default=MLflowDefaults.TRACKING_URI, description="MLflow tracking URI."),
    ]
    experiment: Annotated[
        str,
        Field(default=MLflowDefaults.EXPERIMENT, description="MLflow experiment name."),
    ]
    run_name: Annotated[
        str | None,
        Field(default=MLflowDefaults.RUN_NAME, description="MLflow run name."),
    ]
    tags: Annotated[
        str | None,
        Field(default=MLflowDefaults.TAGS, description="Comma-separated MLflow tags."),
    ]
    parent_run_id: Annotated[
        str | None,
        Field(default=None, description="Optional MLflow parent run ID."),
    ]
    artifact_globs: Annotated[
        list[str] | None,
        Field(
            default=MLflowDefaults.ARTIFACT_GLOBS,
            description="Artifact glob overrides for MLflow upload.",
        ),
    ]

    @property
    def enabled(self) -> bool:
        """Whether MLflow export/live streaming is enabled."""
        return self.tracking_uri is not None

    @property
    def tags_dict(self) -> dict[str, str]:
        """Parse comma-separated ``key:value`` MLflow tags."""
        if not self.tags:
            return {}
        tags: dict[str, str] = {}
        for item in self.tags.split(","):
            key, sep, value = item.partition(":")
            if sep and key.strip() and value.strip():
                tags[key.strip()] = value.strip()
        return tags

    @property
    def resolved_artifact_globs(self) -> list[str]:
        """Return MLflow artifact globs, applying defaults when unset."""
        return list(self.artifact_globs or MLflowDefaults.DEFAULT_ARTIFACT_GLOBS)

File: rust/aiperf/src/export/mlflow.rs
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! MLflow run tracker (native Rust, MLflow REST uploader).
//!
//! Ports the Python MLflow exporter (`exporters/mlflow_data_exporter.py`) to the
//! runner: creates/attaches an MLflow run and logs the same params, per-stat
//! metrics (`metric.tag` for avg, `metric.tag.<stat>` for the rest), tags
//! (`aiperf.version`, `benchmark_id`, user tags), and uploads the artifact
//! bundle — all via the MLflow REST API (`/api/2.0/mlflow/*` + artifact upload)
//! rather than the Python SDK. Parity oracle: the logged param/metric-key/tag set
//! and values must match the Python exporter for an identical run, verified
//! against a local `file://`/http tracking store.
//!
//! Spec §6: run under a hard wall-clock timeout on a short-lived runtime so an
//! unreachable tracking server cannot hang shutdown — the durable requirement the
//! Python subprocess apparatus existed to satisfy, without `spawn`/`Queue`/pickle.
//!
//! STATUS: config + gating are wired; the REST uploader body is unimplemented
//! (Worker C).

use std::path::Path;

use crate::export::{ExportConfig, Exporter};
use crate::metrics_core::NativeReport;

/// MLflow export policy. `enabled` iff a tracking URI is provided (matching the
/// Python `MLflowConfig.enabled`). Worker C extends this struct as needed.
#[derive(Debug, Clone, Default, serde::Deserialize)]
#[serde(default, deny_unknown_fields)]
pub struct MlflowExportConfig {
    /// Whether MLflow tracking is enabled.
    pub enabled: bool,
    /// MLflow tracking URI (`file://…` or `http(s)://…`).
    pub tracking_uri: Option<String>,
    /// Experiment name (Python default `"aiperf"`).
    pub experiment: Option<String>,
    /// Optional run name.
    pub run_name: Option<String>,
    /// Optional parent run id.
    pub parent_run_id: Option<String>,
    /// User tags to attach to the run.
    #[serde(default)]
    pub tags: std::collections::BTreeMap<String, String>,
    /// Artifact globs to upload (relative to the artifact dir).
    #[serde(default)]
    pub artifact_globs: Vec<String>,
}

/// The MLflow REST [`Exporter`].
pub struct MlflowExporter;

impl Exporter for MlflowExporter {
    fn name(&self) -> &'static str {
        "mlflow"
    }

    fn enabled(&self, cfg: &ExportConfig) -> bool {
        cfg.mlflow.enabled && cfg.mlflow.tracking_uri.is_some()
    }

    fn export(
        &self,
        _report: &NativeReport,
        _artifact_dir: &Path,
        _cfg: &ExportConfig,
    ) -> anyhow::Result<()> {
        // Worker C: create/attach the MLflow run and log params/metrics/tags +
        // upload artifacts via REST under a hard timeout. Inert at foundation.
        anyhow::bail!("native MLflow uploader not yet implemented");
    }
}
